src/camera.py

```diff
@@ -0,0 +1,42 @@
+import pygame
+
+from src.gui.scene_animation import SceneAnimation
+from src.settings import SCREEN_HEIGHT, SCREEN_WIDTH
+from src.sprites.base import Sprite
+
+
+class Camera:
+    def __init__(self, width: int, height: int):
+        self._width, self._height = width, height
+        self.state = pygame.Rect(0, 0, width, height)
+
+    def change_size(self, width: int, height: int):
+        self._width, self._height = width, height
+        self.state.size = width, height
+
+    def update(self, target: Sprite | SceneAnimation):
+        self.state.update(self._complex_camera(target.rect))
+
+    def apply(self, target: Sprite):
+        return target.rect.move(self.state.topleft)
+
+    @property
+    def size(self):
+        return self._width, self._height
+
+    def _complex_camera(self, target_rect: pygame.Rect):
+        left, top = target_rect.topleft
+        w, h = self.size
+        left, top = (
+            (SCREEN_WIDTH / 2) - left,
+            (SCREEN_HEIGHT / 2) - top,
+        )  # center player
+
+        left = min(0, left)  # stop scrolling at the left edge
+        left = max(-(w - SCREEN_WIDTH), left)  # stop scrolling at the right edge
+        top = max(-(h - SCREEN_HEIGHT), top)  # stop scrolling at the bottom
+        top = min(0, top)  # stop scrolling at the top
+
+        return pygame.Rect(
+            left - (target_rect.width / 2), top - (target_rect.height / 2), w, h
+        )
```

Comment:
I think the substraction of half the target_rect size from the return rect should be moved to the first min / max statements of the left / top variables to avoid shifting the camera boundaries slightly to the bottom right.
```suggestion
        left = min(0, left - (target_rect.width / 2))  # stop scrolling at the left edge
        left = max(-(w - SCREEN_WIDTH), left)  # stop scrolling at the right edge
        top = max(-(h - SCREEN_HEIGHT), top - (target_rect.height / 2))  # stop scrolling at the bottom
        top = min(0, top)  # stop scrolling at the top

        return pygame.Rect(left, top, w, h)
```
That said, good job! I'm looking forward to implementing this in the mini-game :D

Regarding that, I think there could also be added some kind of zoom / change in the scale factor between two points of a tracking shot, to get a rough overview of large parts of the map at the start of a mini-game or generally when entering a new map for the first time.

Reply:
Actually it would make more sense to just move it to the definition of top/left 😅

Reply:
> Regarding that, I think there could also be added some kind of zoom / change in the scale factor between two points of a tracking shot, to get a rough overview of large parts of the map at the start of a mini-game or generally when entering a new map for the first time.

Oh, I didn't see that! I guess that could be another addition, yes. Also, the camera movement doesn't prevent the player from moving all alone, so should probably do something about that